src/Microsoft.DotNet.CodeFormatter.Analyzers/ExplicitVariableTypeAnalyzer.cs

```diff
@@ -0,0 +1,116 @@
+                // Implicit typed variables cannot have multiple declartors  
+                if (node.Type.IsVar &&
+                    !IsTypeObvious(node) &&
+                    !IsAnonymousType(node.Variables.Single(), syntaxContext.SemanticModel, syntaxContext.CancellationToken))
+                {
+                    syntaxContext.ReportDiagnostic(Diagnostic.Create(s_ruleVariableDeclaration, node.GetLocation(), node.Variables.Single().Identifier.Text));
+                }
+            }, SyntaxKind.VariableDeclaration);
+
+            context.RegisterSyntaxNodeAction(syntaxContext =>
+            {
+                var node = (ForEachStatementSyntax)syntaxContext.Node; 
+                if (node.Type.IsVar &&
+                    !IsTypeObvious(node) &&
+                    !IsAnonymousType(node, syntaxContext.SemanticModel, syntaxContext.CancellationToken))
+                {
+                    syntaxContext.ReportDiagnostic(Diagnostic.Create(s_ruleForEachStatement, node.Identifier.GetLocation(), node.Identifier.Text));
+                }
+            }, SyntaxKind.ForEachStatement);
+        }
+
+        // Return true if given SyntaxNode has an anonymous type, which can't be replaced by an explicit type.
+        private static bool IsAnonymousType(SyntaxNode node, SemanticModel model, CancellationToken cancellationToken)
+        {
+            ISymbol symbol = model.GetDeclaredSymbol(node, cancellationToken); 
+            return ((ILocalSymbol)symbol).Type.IsAnonymousType;
+        }
+
+        // Return true if given SyntaxNode is either VariableDeclarationSyntax or ForEachStatementSyntax and 
+        // VariableDeclarationSyntax.Variables.Single().Initializer.Value or ForEachStatementSyntax.Expression is:
+        //   1. LiteralExpressionSyntax, e.g. var x = 10;
+        //   2. CastExpressionSyntax, e.g. var x = (Foo)f;
+        //   3. BinaryExpressionSyntax with Kind == AsExpression
+        //   4. A object creation syntax node, which (at least) includes:
+        //          - ObjectCreationExpressionSyntax
+        //          - ArrayCreationExpressionSyntax
+        // 
+        //      ImplicitArrayCreationExpressionSyntax: This one is not included. e.g. new[] {}
+        //
+        // TODO: AnonymousObjectCreationExpressionSyntax could be filtered out here as well, maybe we want to do that?
+        //       The trade-off here is the logic we use to check syntax node is not as accurate as a query to SemanticModel, 
+        //       which is (presumbly) much slower
+        private static bool IsTypeObvious(SyntaxNode node)
+        {
+            
+            if (node == null)
+            {
+                return false;
+            }
+
+            ExpressionSyntax expressionNode = null;
+            if (node is VariableDeclarationSyntax)
+            {
+                expressionNode = ((VariableDeclarationSyntax)node).Variables.Single().Initializer.Value;
```

Comment:
FxCop would complain about the inefficiency of casting twice (once in the `is` expression, once with the explicit cast. Instead:
````
var variableDeclaration = node as VariableDeclarationSyntax;
if (variableDeclaration != null)
{
    expressionNode = variableDeclaration.Variables.Single().Initializer.Value;
    ...
````
````